src/WCSimAmBeGen.cc

```diff
@@ -0,0 +1,110 @@
+#include "WCSimAmBeGen.hh"
+#include "WCSimPrimaryGeneratorMessenger.hh"
+#include "WCSimPrimaryGeneratorAction.hh"
+
+#include "G4ParticleGun.hh"
+#include "G4PhysicalConstants.hh"
+#include "G4RandomDirection.hh"
+#include "G4SystemOfUnits.hh"
+#include "G4SPSEneDistribution.hh"
+#include "G4SPSRandomGenerator.hh"
+#include "G4ParticleTable.hh"
+#include "G4Neutron.hh"
+#include "G4Gamma.hh"
+
+#include <random>
+
+using namespace std;
+
+WCSimAmBeGen::WCSimAmBeGen(){
+  // Initialise
+  this->Initialise();
+
+}
+
+WCSimAmBeGen::~WCSimAmBeGen(){
+  
+  // This needed to be deleted
+  delete myAmBeGun;
+  delete nEnergyDist;
+  delete rGen;
+}
+
+void WCSimAmBeGen::Initialise(){
+    nEnergyDist = new G4SPSEneDistribution();
+    rGen        = new G4SPSRandomGenerator();
+    myAmBeGun   = new G4ParticleGun();
+}
+
+G4double WCSimAmBeGen::GammaEnergy(){
+    
+    // Define the correspondent probabilities for every possible gamma scenario
+    probabilities = {0.26, 0.65, 0.08};
+    energies      = {0.0, 4.4, 7.7};
+
+    // Generate a weighted random number so we can select the gamma energy
+    mt19937 gen(std::random_device{}());
```

Comment:
any reason why you don't use the geant4 built-in random number generator?

Reply:
Not at all, I reworked the code to look like this:

```
G4double WCSimAmBeGen::GammaEnergy(){
    
    // Initialise Geant4 random number generator
    G4Random::setTheEngine(new CLHEP::RanecuEngine);

    // Define arrays
    G4double probabilities[] = {0.26, 0.65, 0.08};
    G4double energies[] = {0.0, 4.4 * MeV, 7.7 * MeV};

    // Generation of the random number
    G4double random_number = G4UniformRand();

    // Select the energy taking into account the probabilities
    G4double cumulative_probability = 0.0;
    G4int selected_energy_index = -1;

    for (G4int j = 0; j < sizeof(probabilities) / sizeof(probabilities[0]); ++j) {
        cumulative_probability += probabilities[j];
        if (random_number <= cumulative_probability) {
            selected_energy_index = j;
            break;
        }
    }

    gEnergy = energies[selected_energy_index];

    return gEnergy; 
}
```